sqlglot/dialects/spark2.py

```diff
@@ -110,10 +115,41 @@ def temporary_storage_provider(expression: exp.Expression) -> exp.Expression:
     return expression
 
 
+def _annotate_by_same_args(
+    self: TypeAnnotator, expression: E, *args: str, target_type: exp.DataType | exp.DataType.Type
+) -> E:
+    """
+    Infers the type of the expression if all the param @args are of that type,
+    otherwise defaults to param @target_type
+    """
+    self._annotate_args(expression)
+
+    expressions: t.List[exp.Expression] = []
+    for arg in args:
+        arg_expr = expression.args.get(arg)
+        expressions.extend(expr for expr in ensure_list(arg_expr) if expr)
+
+    last_datatype = expressions[0].type if expressions else None
+
+    for expr in expressions:
+        if not expr.is_type(last_datatype):
+            last_datatype = None
+            break
+
+    self._set_type(expression, last_datatype or target_type)
+    return expression
```

Comment:
What if you have unknowns? This would preemptively infer the type to be `TEXT` for `CONCAT` even if you had `UNKNOWN` args.

Reply:
That should be fine I think, in the majority of the cases that'll be the result type anyways.

Reply:
If the last type is `BINARY` and the rest are `UNKNOWN`, you then assume `TEXT` even though `BINARY` is also possible (e.g. if the unknown types were `BINARY`).